In this Notebook I intend to perform basic data exploration of the databases provided by the MinTIC for the DS4A Final project.

I already made copies of each table in CSV format and manipulated them a little as to have unique columns plus an additional column of year. This way, I think, I may be able to further manipulate and explore them in Pandas

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
mdm = pd.read_csv("MDM_2016_2019.csv")
comp_dim = pd.read_csv("Componentes_Dimensiones_2016_2019.csv", header=1)
investments = pd.read_csv("BD_MinTIC_2010_2020.csv")
pd.options.display.max_columns = 100

First we are interested in looking for the basic structure of each table, beginning by the MDM table:

In [4]:
mdm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4404 entries, 0 to 4403
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Anio_corte                   4404 non-null   int64  
 1   Municipio                    4404 non-null   object 
 2   Código                       4404 non-null   int64  
 3   Departamento                 4404 non-null   object 
 4   Grupo_dotaciones             4404 non-null   object 
 5   Categoría_de_ruralidad       4404 non-null   object 
 6   MDM                          4404 non-null   float64
 7   Puesto_MDM_Grupo_Dotaciones  4404 non-null   int64  
 8   Puesto_MDM_Nacional          4404 non-null   int64  
dtypes: float64(1), int64(4), object(4)
memory usage: 309.8+ KB


The dtype of each column is as expected.

So, next we will examine the numeric variables of this table, looking for its minimum and maximum values, as well as a basic grasp on its distributions. Though we won't treat the year nor the municipality codes as numeric values in further analyses, it is still interesting to see which are the minimum and maximum values these varibles take.

In [5]:
mdm.describe()

Anio_corte        Código          MDM  Puesto_MDM_Grupo_Dotaciones  \
count  4404.000000   4404.000000  4404.000000                  4404.000000   
mean   2017.500000  37671.950954    50.372279                   108.092643   
std       1.118161  25705.991227    10.170328                    63.422778   
min    2016.000000   5001.000000     2.033311                     1.000000   
25%    2016.750000  15646.000000    43.950000                    53.000000   
50%    2017.500000  25781.000000    50.209631                   108.000000   
75%    2018.250000  63302.000000    56.550113                   163.000000   
max    2019.000000  99773.000000    86.816932                   218.000000   

       Puesto_MDM_Nacional  
count          4404.000000  
mean            551.000000  
std             317.867283  
min               1.000000  
25%             276.000000  
50%             551.000000  
75%             826.000000  
max            1101.000000

Finally, we are interested in creating a synthetic primary key for this table, in order to merge its information to the other tables in further analyses:

In [13]:
mdm["id"] = mdm["Anio_corte"].astype("string")+mdm["Código"].astype("string")
mdm

Anio_corte       Municipio  Código Departamento Grupo_dotaciones  \
0           2016        MEDELLÍN    5001    ANTIOQUIA                C   
1           2016       ABEJORRAL    5002    ANTIOQUIA               G3   
2           2016        ABRIAQUÍ    5004    ANTIOQUIA               G3   
3           2016      ALEJANDRÍA    5021    ANTIOQUIA               G2   
4           2016           AMAGÁ    5030    ANTIOQUIA               G2   
...          ...             ...     ...          ...              ...   
4399        2019         TARAIRA   97666       VAUPES               G2   
4400        2019  PUERTO_CARREÑO   99001      VICHADA               G1   
4401        2019    LA_PRIMAVERA   99524      VICHADA               G4   
4402        2019   SANTA_ROSALÍA   99624      VICHADA               G2   
4403        2019        CUMARIBO   99773      VICHADA               G5   

         Categoría_de_ruralidad        MDM  Puesto_MDM_Grupo_Dotaciones  \
0     Ciudades_y_aglomeraciones  80.780000                            1   
1                         Rural  49.480000                           76   
2                Rural_disperso  44.940000                          132   
3                         Rural  54.940000                           66   
4                   Intermedios  51.880000                           98   
...                         ...        ...                          ...   
4399             Rural_disperso  42.470492                          207   
4400             Rural_disperso  57.566808                          167   
4401             Rural_disperso  56.287012                           60   
4402             Rural_disperso  48.886473                          188   
4403             Rural_disperso  41.512946                          168   

      Puesto_MDM_Nacional         id  
0                       3   20165001  
1                     445   20165002  
2                     678   20165004  
3                     232   20165021  
4                     334   20165030  
...                   ...        ...  
4399                  978  201997666  
4400                  425  201999001  
4401                  492  201999524  
4402                  825  201999624  
4403                  994  201999773  

[4404 rows x 10 columns]

Now we will repeat the same steps for the MinTIC investments table:

In [38]:
investments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63361 entries, 0 to 63360
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   ANIO_CORTE             63361 non-null  int64 
 1   FECHA_CORTE            63361 non-null  object
 2   DEPARTAME_NOMBRE       63361 non-null  object
 3   DEPARTAME_COD          63361 non-null  object
 4   MUNICIPIO_NOMBRE       63361 non-null  object
 5   MUNICIPIO_COD          63361 non-null  int64 
 6   INICIATIVA             63361 non-null  object
 7   VICEMINISTERIO         63361 non-null  object
 8   INDICADOR              63361 non-null  object
 9   DESCRIPCION_INDICADOR  63361 non-null  object
 10  INVERSION              63361 non-null  int64 
 11  BENEFICIARIOS          63361 non-null  int64 
dtypes: int64(4), object(8)
memory usage: 5.8+ MB


As before, the column dtypes are as expected, except for the `DEPARTAME_COD` column, which appears as an `object` dtype, while we expected an integer dtype. Also the `FECHA_CORTE` column should be a `data` dtype, but it is an `object`. Let's examine first the `DEPARTAME_COD` column:

In [40]:
investments["DEPARTAME_COD"].unique()

array(['11', '5', '76', '25', '8', '52', '70', '13', '73', '15', '66',
       '99', '68', '23', '63', '41', '44', '94', '91', '17', '19', '54',
       '95', '20', '50', '18', '27', '86', '47', '81', '97', '85', '88',
       '0', 'IN'], dtype=object)

Ok, so the reason it is treated as an `object` dtype is that 'IN' value.

In [42]:
investments[investments["DEPARTAME_COD"]=="IN"]

ANIO_CORTE FECHA_CORTE DEPARTAME_NOMBRE DEPARTAME_COD MUNICIPIO_NOMBRE  \
17655        2020  01/20/2020    INTERNACIONAL            IN    INTERNACIONAL   
42812        2020  12/31/2020    INTERNACIONAL            IN    INTERNACIONAL   
60182        2020  01/20/2020    INTERNACIONAL            IN    INTERNACIONAL   

       MUNICIPIO_COD                              INICIATIVA  VICEMINISTERIO  \
17655           1123  HABILIDADES_DIGITALES_CIENCIA_DE_DATOS  Transformacion   
42812           1123                      CIUDADANIA_DIGITAL  Transformacion   
60182           1123  HABILIDADES_DIGITALES_CIENCIA_DE_DATOS  Transformacion   

                                             INDICADOR  \
17655         Colombianos_formados_en_ciencia_de_datos   
42812  Número_de_formaciones_en_competencias_digitales   
60182         Colombianos_formados_en_ciencia_de_datos   

                                   DESCRIPCION_INDICADOR  INVERSION  \
17655           Colombianos_formados_en_ciencia_de_datos   16118564   
42812  Se_mide_el_avance_de_las_certificaciones_entre...          0   
60182           Colombianos_formados_en_ciencia_de_datos          0   

       BENEFICIARIOS  
17655              4  
42812            408  
60182             14

We see now that it is indeed a special code used for International investment projects. Perhaps it would be a good idea to try and think whether it would be a good idea to replace this code with a particular number (as to treat the entire column as an integer) or to let it as is and treat it as an `object` dtype. 

Next we will examine the numeric variables (again, as before):

In [43]:
investments.describe()

ANIO_CORTE  MUNICIPIO_COD     INVERSION  BENEFICIARIOS
count  63361.000000   63361.000000  6.336100e+04   6.336100e+04
mean    2017.484352   38704.559051  1.936421e+08   9.054179e+02
std        3.145821   26396.793814  3.279701e+09   3.508134e+04
min     1990.000000       0.000000  0.000000e+00   0.000000e+00
25%     2015.000000   15676.000000  0.000000e+00   0.000000e+00
50%     2019.000000   25839.000000  0.000000e+00   1.000000e+00
75%     2020.000000   66572.000000  2.568316e+07   2.500000e+01
max     2021.000000   99773.000000  7.735770e+11   7.592871e+06

In [44]:
print(f'Número de municipios registrados en MDM: {len(mdm["Código"].unique())}')
print(f'Número de municipios registrados en BD_MinTIC: {len(investments["MUNICIPIO_COD"].unique())}')

Número de municipios registrados en MDM: 1101
Número de municipios registrados en BD_MinTIC: 1146


According to the `describe` results of the MinTIC investments table, we can see that there are municipality codes that doesn't appear on the MDM table. We can see this because in the MDM table the minimum code was 5001, while in the investments table the minimum value is 0. This is further corroborated examining the number of codes appearing in each table, with a difference of 45 between them. In the next cell we look for the differing municipalities:

In [45]:
investments[~investments["MUNICIPIO_COD"].isin(mdm["Código"])].dropna()[["DEPARTAME_NOMBRE","MUNICIPIO_NOMBRE"]].drop_duplicates().sort_values("DEPARTAME_NOMBRE")

DEPARTAME_NOMBRE  MUNICIPIO_NOMBRE
46                     AMAZONAS          TARAPACA
82                     AMAZONAS      PUERTO_ARICA
115                    AMAZONAS    PUERTO_ALEGRIA
116                    AMAZONAS        EL_ENCANTO
7097                   AMAZONAS       LA_VICTORIA
748                    AMAZONAS       LA_CHORRERA
7079                   AMAZONAS  PUERTO_SANTANDER
3859                   AMAZONAS        LA_PEDRERA
4919                   AMAZONAS   MIRITI_-_PARANA
46298                 ANTIOQUIA       GOBERNACION
46164                    ARAUCA       GOBERNACION
46678                 ATLANTICO       GOBERNACION
46219                   BOLIVAR       GOBERNACION
46196                   CAQUETA       GOBERNACION
46799                     CESAR       GOBERNACION
7148                      CHOCO   BELÉN_DE_BAJIRA
46538                     CHOCO       GOBERNACION
46174              CUNDINAMARCA       GOBERNACION
165                     GUAINIA   PUERTO_COLOMBIA
4968                    GUAINIA         PANA_PANA
133                     GUAINIA    BARRANCO_MINAS
747                     GUAINIA      LA_GUADALUPE
15338                   GUAINIA          MORICHAL
46457                   GUAINIA       GOBERNACION
163                     GUAINIA          CACAHUAL
164                     GUAINIA        SAN_FELIPE
4872                    GUAINIA        MAPIRIPANA
46201                  GUAVIARE       GOBERNACION
17655             INTERNACIONAL     INTERNACIONAL
46525                 MAGDALENA       GOBERNACION
15895                  NACIONAL          NACIONAL
46172                    NARIÑO       GOBERNACION
46169        NORTE_DE_SANTANDER       GOBERNACION
46345                  PUTUMAYO       GOBERNACION
46391                   QUINDIO       GOBERNACION
46694                 RISARALDA       GOBERNACION
46193                 SANTANDER       GOBERNACION
456    SAN_ANDRES_Y_PROVIDENCIA        SAN_ANDRES
46209                    TOLIMA       GOBERNACION
46616           VALLE_DEL_CAUCA       GOBERNACION
28669                    VAUPES          PAPUNAUA
46571                    VAUPES       GOBERNACION
1370                     VAUPES             PACOA
3900                     VAUPES          YAVARATE
46423                   VICHADA       GOBERNACION

Finally, again we are interested in creating a synthetic primary key for the investment table, formed by the year and the municipality code:

In [46]:
investments["id"] = investments["ANIO_CORTE"].astype("string")+investments["MUNICIPIO_COD"].astype("string")
investments

ANIO_CORTE FECHA_CORTE DEPARTAME_NOMBRE DEPARTAME_COD  \
0            2020  12/31/2020           BOGOTA            11   
1            2020  12/31/2020        ANTIOQUIA             5   
2            2020  12/31/2020        ANTIOQUIA             5   
3            2020  12/31/2020  VALLE_DEL_CAUCA            76   
4            2020  12/31/2020     CUNDINAMARCA            25   
...           ...         ...              ...           ...   
63356        2021  02/12/2021         PUTUMAYO            86   
63357        2021  02/12/2021         PUTUMAYO            86   
63358        2021  02/12/2021         PUTUMAYO            86   
63359        2021  02/12/2021         GUAVIARE            95   
63360        2021  02/12/2021           VAUPES            97   

            MUNICIPIO_NOMBRE  MUNICIPIO_COD                INICIATIVA  \
0                BOGOTA_D.C.          11001  ASESORIAS_EN_TELETRABAJO   
1                   MEDELLIN           5001  ASESORIAS_EN_TELETRABAJO   
2                   APARTADO           5045  ASESORIAS_EN_TELETRABAJO   
3                       CALI          76001  ASESORIAS_EN_TELETRABAJO   
4                VILLAPINZON          25873  ASESORIAS_EN_TELETRABAJO   
...                      ...            ...                       ...   
63356              LEGUIZAMO          86573            VENDE_EN_LÍNEA   
63357          SAN_FRANCISCO          86755            VENDE_EN_LÍNEA   
63358               SANTIAGO          86760            VENDE_EN_LÍNEA   
63359  SAN_JOSE_DEL_GUAVIARE          95001            VENDE_EN_LÍNEA   
63360                 CARURU          97161            VENDE_EN_LÍNEA   

       VICEMINISTERIO                                          INDICADOR  \
0      Transformacion  Número_de_personas_impactadas_con_la_difusión_...   
1      Transformacion  Número_de_personas_impactadas_con_la_difusión_...   
2      Transformacion  Número_de_personas_impactadas_con_la_difusión_...   
3      Transformacion  Número_de_personas_impactadas_con_la_difusión_...   
4      Transformacion  Número_de_personas_impactadas_con_la_difusión_...   
...               ...                                                ...   
63356  Transformacion                                        No_asociado   
63357  Transformacion                                        No_asociado   
63358  Transformacion                                        No_asociado   
63359  Transformacion                                        No_asociado   
63360  Transformacion                                        No_asociado   

                                   DESCRIPCION_INDICADOR     INVERSION  \
0      Número_de_personas_impactadas_con_la_difusión_...  773577049710   
1      Número_de_personas_impactadas_con_la_difusión_...  114887680650   
2      Número_de_personas_impactadas_con_la_difusión_...   70209138175   
3      Número_de_personas_impactadas_con_la_difusión_...   59996899895   
4      Número_de_personas_impactadas_con_la_difusión_...   47231602045   
...                                                  ...           ...   
63356                                        No_asociado             0   
63357                                        No_asociado             0   
63358                                        No_asociado             0   
63359                                        No_asociado             0   
63360                                        No_asociado             0   

       BENEFICIARIOS         id  
0                606  202011001  
1                 90   20205001  
2                 55   20205045  
3                 47  202076001  
4                 37  202025873  
...              ...        ...  
63356              1  202186573  
63357              1  202186755  
63358              3  202186760  
63359             24  202195001  
63360              1  202197161  

[63361 rows x 13 columns]

Now we will examine the last table, Components Dimensions:

In [47]:
comp_dim.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4404 entries, 0 to 4403
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   Anio_corte                           4404 non-null   int64  
 1   Municipio                            4404 non-null   object 
 2   Código                               4404 non-null   int64  
 3   Departamento                         4404 non-null   object 
 4   Grupo_dotaciones                     4404 non-null   object 
 5   Categoría_de_ruralidad               4404 non-null   object 
 6   Cobertura_media_neta                 4404 non-null   float64
 7   SABER_11_Matematicas                 4404 non-null   float64
 8   SABER_11_Lenguaje                    4404 non-null   float64
 9   Cobertura_Transición                 4404 non-null   float64
 10  Cobertura_salud                      4404 non-null   float64
 11  Vacunación_Pentavalente       

Anything unusual here in the dtypes of the columns, except for the rural electric coverage variable, which is an `object` dtype. Let's examine this variable:

In [51]:
comp_dim["Cobertura_eléctrica_rural"].unique()

array(['1', '0.98', '0.95', '0.97', '0.99', '0.88', '0.54', '0.91', '0.9',
       '0.94', '0.7', '0.85', '0.96', '0.89', '0.77', '0.78', '0.93',
       '0.68', '0.84', '0.79', '0.71', '0.76', '0.81', '0.67', '0.92',
       '0.8', '0.64', '0.82', '0.43', '0.55', '0.25', '0.52', '0.75',
       '0.83', '0.61', '0.35', '0.42', '0.74', '0.6', '0.65', '0.19',
       '0.86', '0.26', '0.63', '0.44', '0.46', '0.69', '0.48', '0.2',
       '0.09', '_-______', '0.87', '0.58', '0.5', '0.45', '0.4', '0.41',
       '0.13', '0.08', '0.56', '0.73', '0.72', '0.38', '0.24', '0.51',
       '0.62', '0.53', '0.15', '0.59', '0.29', '0.39', '0.37', '0.16',
       '0.66', '0.57', '0.1', '0.47', '0.02', '0.49', '0.11', '0.06',
       '0.3', '0.33', '0.28', '0.22', '0.36', '0.14', '0.27', '0.21',
       '0.17', '0.03', '0.31', '0', '0.07', '0.04', '0.34', '0.23', nan,
       '0.18', '0.971', '0.851', '0.984', '0.972', '0.983', '0.976',
       '0.743', '0.966', '0.941', '0.985', '0.853', '0.973', '0.982',
       

We already see there are an interesting element '_-______' which cannot be converted to numeric whatsoever. This element surely represent missing data, and has this particular form since in the basic cleaning process I removed all the blanks and replaced them with underscores `_`. Let's see which rows have this element in the column: 

In [52]:
nonfloats=[]
for i in range(len(comp_dim)):
    try:
        float(comp_dim["Cobertura_eléctrica_rural"][i])
    except ValueError:
        nonfloats.append(i)
print(nonfloats)
print(comp_dim[["Anio_corte","Municipio","Departamento","Cobertura_eléctrica_rural"]].loc[nonfloats])

[267, 456, 599, 663, 1086]
      Anio_corte            Municipio Departamento Cobertura_eléctrica_rural
267         2016                PISBA       BOYACA                  _-______
456         2016               TUCHÍN      CORDOBA                  _-______
599         2016  SAN_JOSÉ_DEL_PALMAR        CHOCO                  _-______
663         2016            CONCORDIA    MAGDALENA                  _-______
1086        2016          PROVIDENCIA   SAN_ANDRES                  _-______


And let's get rid of them (by replacing them with na's):

In [53]:
cobertura_electrica_floats=[]
for i in range(len(comp_dim)):
    try:
        cobertura_electrica_floats.append(float(comp_dim["Cobertura_eléctrica_rural"][i]))
    except ValueError:
        cobertura_electrica_floats.append(np.nan)
comp_dim["Cobertura_eléctrica_rural"] = cobertura_electrica_floats

Let's check again and see whether the problem was fixed:

In [54]:
comp_dim.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4404 entries, 0 to 4403
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   Anio_corte                           4404 non-null   int64  
 1   Municipio                            4404 non-null   object 
 2   Código                               4404 non-null   int64  
 3   Departamento                         4404 non-null   object 
 4   Grupo_dotaciones                     4404 non-null   object 
 5   Categoría_de_ruralidad               4404 non-null   object 
 6   Cobertura_media_neta                 4404 non-null   float64
 7   SABER_11_Matematicas                 4404 non-null   float64
 8   SABER_11_Lenguaje                    4404 non-null   float64
 9   Cobertura_Transición                 4404 non-null   float64
 10  Cobertura_salud                      4404 non-null   float64
 11  Vacunación_Pentavalente       

Alright! So now let's describe those numeric variables:

In [12]:
(comp_dim["Anio_corte"].unique())

array([2016, 2017, 2018, 2019])

In [55]:
comp_dim.describe()

Anio_corte        Código  Cobertura_media_neta  SABER_11_Matematicas  \
count  4404.000000   4404.000000           4404.000000           4404.000000   
mean   2017.500000  37671.950954              0.407100             47.950800   
std       1.118161  25705.991227              0.156045              4.823078   
min    2016.000000   5001.000000              0.013100              0.000000   
25%    2016.750000  15646.000000              0.307000             44.750000   
50%    2017.500000  25781.000000              0.404900             48.195500   
75%    2018.250000  63302.000000              0.497000             51.356250   
max    2019.000000  99773.000000              1.000000             66.130000   

       SABER_11_Lenguaje  Cobertura_Transición  Cobertura_salud  \
count        4404.000000           4404.000000      4404.000000   
mean           49.880186              0.535216         0.808769   
std             3.592606              0.164099         0.160228   
min             0.000000              0.051000         0.170000   
25%            47.703000              0.423000         0.710000   
50%            50.129000              0.532050         0.828000   
75%            52.351750              0.637000         0.945000   
max            64.130000              1.000000         1.000000   

       Vacunación_Pentavalente  Mortalidad_Infantil  \
count              4404.000000          4404.000000   
mean                 45.912805            18.863228   
std                  46.425937             9.315656   
min                   0.170000             1.900000   
25%                   0.950000            13.510000   
50%                   8.350000            17.160000   
75%                  92.200000            21.914000   
max                 339.400000            87.910000   

       Cobertura_eléctrica_rural  Cobertura_internet  Cobertura_Acueducto  \
count                4398.000000         4404.000000          4404.000000   
mean                    0.879465            1.775996             0.571314   
std                     0.188705            3.925243             0.298542   
min                     0.000000            0.000000             0.000000   
25%                     0.860000            0.014000             0.340000   
50%                     0.960000            0.100000             0.570000   
75%                     0.990000            1.800000             0.832000   
max                     1.000000           68.600000             1.000000   

       Cobertura_Alcantarillado         Hurtos  Hurtos_x_10000_hab  \
count               4404.000000    4404.000000         4404.000000   
mean                   0.388369     254.149637           17.872695   
std                    0.281378    3363.223109           21.107681   
min                    0.000000       0.000000            0.000000   
25%                    0.160000       4.000000            4.460000   
50%                    0.340000      12.000000           10.560000   
75%                    0.580000      44.000000           22.990000   
max                    1.000000  156435.000000          206.028787   

        Homicidios  Homicidios_x_10000_hab  Violencia_intrafamiliar  \
count  4404.000000             4404.000000              4404.000000   
mean     11.413261                2.480690                80.103315   
std      57.512463                3.171105               934.408670   
min       0.000000                0.000000                 0.000000   
25%       0.000000                0.000000                 3.000000   
50%       2.000000                1.590000                 9.000000   
75%       7.250000                3.273055                23.000000   
max    1288.000000               31.630000             36179.000000   

       Violencia_intrafamiliar_x_10000_hab  
count                          4404.000000  
mean                             10.004996  
std                              10.011659  
min                               0.000000

Here the main theme to look at are those variables described as "coverages", since we already saw that some of these variables are in two forms: as percentage (0%-100%) as well as proportions (0-1), depending on the year. So we should convert these variables to the same form (either percentage or proportion). Here I choose to convert all of them to proportions since I think it will be easier ad more confortable to do math with them in this form. 

First we take all the variables described as "coverages":

In [56]:
variables = list(comp_dim.columns[comp_dim.columns.str.contains("cobertura", case=False)])+['Vacunación_Pentavalente']

Then we try and convert every value above 1 to its corresponding value between 0 and 1 (we'll surely have to double-check this procedure with the original dataset):

In [57]:
for variable in variables:
    comp_dim[variable] = [comp_dim[variable][i]/100 if comp_dim[variable][i] > 1 else comp_dim[variable][i] for i in range(len(comp_dim))]

comp_dim[variables].describe()

Cobertura_media_neta  Cobertura_Transición  Cobertura_salud  \
count           4404.000000           4404.000000      4404.000000   
mean               0.407100              0.535216         0.808769   
std                0.156045              0.164099         0.160228   
min                0.013100              0.051000         0.170000   
25%                0.307000              0.423000         0.710000   
50%                0.404900              0.532050         0.828000   
75%                0.497000              0.637000         0.945000   
max                1.000000              1.000000         1.000000   

       Cobertura_eléctrica_rural  Cobertura_internet  Cobertura_Acueducto  \
count                4398.000000         4404.000000          4404.000000   
mean                    0.879465            0.101718             0.571314   
std                     0.188705            0.193781             0.298542   
min                     0.000000            0.000000             0.000000   
25%                     0.860000            0.012000             0.340000   
50%                     0.960000            0.030000             0.570000   
75%                     0.990000            0.097250             0.832000   
max                     1.000000            1.000000             1.000000   

       Cobertura_Alcantarillado  Vacunación_Pentavalente  
count               4404.000000              4404.000000  
mean                   0.388369                 0.910899  
std                    0.281378                 0.137858  
min                    0.000000                 0.157000  
25%                    0.160000                 0.846000  
50%                    0.340000                 0.938000  
75%                    0.580000                 1.000000  
max                    1.000000                 3.394000

All seem good and sound. However, I did not see before that the `Vacunación_Pentavalente` column (which is described in the data dictionary as a percentage) has originally values above 100% (339% indeed!). We should examine which rows have values above 1 for that variable:

In [61]:
comp_dim[comp_dim["Vacunación_Pentavalente"]>1]

Anio_corte    Municipio  Código Departamento Grupo_dotaciones  \
1107        2017        ANDES    5034    ANTIOQUIA               G3   
1108        2017  ANGELÓPOLIS    5036    ANTIOQUIA               G3   
1109        2017    ANGOSTURA    5038    ANTIOQUIA               G3   
1112        2017         ANZA    5044    ANTIOQUIA               G4   
1119        2017        BELLO    5088    ANTIOQUIA               G1   
...          ...          ...     ...          ...              ...   
2172        2017     TRINIDAD   85430     CASANARE               G1   
2175        2017        COLÓN   86219     PUTUMAYO               G3   
2184        2017     SANTIAGO   86760     PUTUMAYO               G5   
2186        2017  VILLAGARZÓN   86885     PUTUMAYO               G1   
2190        2017      INÍRIDA   94001      GUAINIA               G2   

         Categoría_de_ruralidad  Cobertura_media_neta  SABER_11_Matematicas  \
1107                Intermedios                0.3340                 47.48   
1108                Intermedios                0.3141                 43.90   
1109             Rural_disperso                0.4928                 41.28   
1112             Rural_disperso                0.1895                 45.61   
1119  Ciudades_y_aglomeraciones                0.3829                 49.35   
...                         ...                   ...                   ...   
2172             Rural_disperso                0.2709                 49.24   
2175                Intermedios                0.3990                 52.62   
2184                      Rural                0.2454                 47.98   
2186                      Rural                0.4337                 48.23   
2190             Rural_disperso                0.1643                 50.01   

      SABER_11_Lenguaje  Cobertura_Transición  Cobertura_salud  \
1107              51.66                0.5431            0.820   
1108              50.00                0.2966            0.473   
1109              46.09                0.4074            0.906   
1112              49.00                0.3529            0.895   
1119              53.49                0.5337            0.849   
...                 ...                   ...              ...   
2172              50.96                0.6080            0.733   
2175              54.25                0.4100            0.869   
2184              50.57                0.2837            0.682   
2186              50.98                0.5165            1.000   
2190              53.42                0.8129            1.000   

      Vacunación_Pentavalente  Mortalidad_Infantil  Cobertura_eléctrica_rural  \
1107                    1.012                14.45                       0.98   
1108                    1.017                13.51                       0.98   
1109                    1.192                27.41                       0.97   
1112                    1.130                14.93                       0.93   
1119                    1.019                 8.77                       1.00   
...                       ...                  ...                        ...   
2172                    1.004                29.83                       0.55   
2175                    1.113                16.30                       0.86   
2184                    1.319                16.96                       0.56   
2186                    1.064                27.79                       0.61   
2190                    1.047                30.30                       1.00   

      Cobertura_internet  Cobertura_Acueducto  Cobertura_Alcantarillado  \
1107               0.061                 0.51                      0.41   
1108               0.022                 0.79                      0.77   
1109               0.027                 0.54                      0.10   
1112               0.020                 0.21                      0.14   
1119               0.207                 1.00                  

285 rows have indeed proportion values above 1, which should be impossible.

In [65]:
print(f'the year(s) that these anomalous values appear are: {comp_dim[comp_dim["Vacunación_Pentavalente"]>1]["Anio_corte"].unique()}')
print(f'The proportion of 2017 registers that these represent is {len(comp_dim[comp_dim["Vacunación_Pentavalente"]>1])/len(comp_dim[comp_dim["Anio_corte"]==2017])}')

the year(s) that these anomalous values appear are: [2017]
The proportion of 2017 registers that these represent is 0.25885558583106266


An interesting feature of those registers is that all of them are from the 2017 year. Also, they represent over 25% of the 2017 data, so we cannot dismiss them easily. Well, I think we will have to think also about this.

Finally, let's create the primary key for this table: 

In [66]:
comp_dim["id"] = comp_dim["Anio_corte"].astype("string")+comp_dim["Código"].astype("string")
comp_dim

Anio_corte       Municipio  Código Departamento Grupo_dotaciones  \
0           2016        MEDELLÍN    5001    ANTIOQUIA                C   
1           2016       ABEJORRAL    5002    ANTIOQUIA               G3   
2           2016        ABRIAQUÍ    5004    ANTIOQUIA               G3   
3           2016      ALEJANDRÍA    5021    ANTIOQUIA               G2   
4           2016           AMAGÁ    5030    ANTIOQUIA               G2   
...          ...             ...     ...          ...              ...   
4399        2019         TARAIRA   97666       VAUPES               G2   
4400        2019  PUERTO_CARREÑO   99001      VICHADA               G1   
4401        2019    LA_PRIMAVERA   99524      VICHADA               G4   
4402        2019   SANTA_ROSALÍA   99624      VICHADA               G2   
4403        2019        CUMARIBO   99773      VICHADA               G5   

         Categoría_de_ruralidad  Cobertura_media_neta  SABER_11_Matematicas  \
0     Ciudades_y_aglomeraciones                 0.559                49.260   
1                         Rural                 0.329                48.100   
2                Rural_disperso                 0.522                46.940   
3                         Rural                 0.330                47.200   
4                   Intermedios                 0.421                45.750   
...                         ...                   ...                   ...   
4399             Rural_disperso                 0.083                45.333   
4400             Rural_disperso                 0.260                46.873   
4401             Rural_disperso                 0.296                49.623   
4402             Rural_disperso                 0.177                44.700   
4403             Rural_disperso                 0.024                42.233   

      SABER_11_Lenguaje  Cobertura_Transición  Cobertura_salud  \
0                51.700                 0.737            1.000   
1                51.280                 0.590            0.882   
2                50.710                 0.848            0.996   
3                49.890                 0.979            1.000   
4                49.870                 0.401            0.788   
...                 ...                   ...              ...   
4399             47.095                 0.286            0.562   
4400             49.463                 0.690            1.000   
4401             51.142                 0.482            0.918   
4402             46.675                 0.317            0.818   
4403             42.813                 0.285            0.539   

      Vacunación_Pentavalente  Mortalidad_Infantil  Cobertura_eléctrica_rural  \
0                       0.971               11.120                      1.000   
1                       0.813               13.160                      0.980   
2                       0.852               16.950                      0.950   
3                       0.831               16.790                      0.970   
4                       0.979               12.560                      0.990   
...                       ...                  ...                        ...   
4399                    0.600               85.616                      0.126   
4400                    1.000               35.981                      0.324   
4401                    0.815               36.501                      0.000   
4402                    0.988               24.618                      0.000   
4403                    0.484               66.887                      0.014   

      Cobertura_internet  Cobertura_Acueducto  Cobertura_Alcantarillado  \
0                  0.214                0.970                     0.920   
1                  0.046                0.370                     0.280   
2                  0.030                0.290                     0.290   
3                  0.079                0.500                     0.310   
4                  0.074   